### PR TITLE
P4 device

### PR DESCRIPTION
**Pull request by pbhide**
Trying (playing with) pull requests

### CODE REVIEW

File: hw/net/rocker/rocker.c [hunk 1]
```diff
@@ -435,6 +438,7 @@ static int cmd_consume(Rocker *r, DescInfo *info)
     cmd = rocker_tlv_get_le16(tlvs[ROCKER_TLV_CMD_TYPE]);
     info_tlv = tlvs[ROCKER_TLV_CMD_INFO];
 
+
```
Comment (scottfeldman) [inline hw/net/rocker/rocker.c]: added whitespce

File: hw/net/rocker/rocker_hw.h [hunk 2]
```diff
@@ -146,6 +147,7 @@ typedef struct rocker_tlv {
 enum {
     ROCKER_TLV_CMD_UNSPEC,
     ROCKER_TLV_CMD_TYPE,                /* u16 */
+    ROCKER_TLV_CMD_WORLD,               /* u32 - world id */
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_hw.h:150]: Since the ROCKER_TLV_CMD_TYPE_P4_xxx cmds are already only for P4 world, it seems we don't need this ROCKER_TLV_CMD_WORLD.  In other words, the world is encoded in the cmd type.

File: hw/net/rocker/rocker_hw.h [hunk 3]
```diff
@@ -165,6 +167,14 @@ enum {
     ROCKER_TLV_CMD_TYPE_OF_DPA_GROUP_DEL,
     ROCKER_TLV_CMD_TYPE_OF_DPA_GROUP_GET_STATS,
 
+    ROCKER_TLV_CMD_TYPE_CLEAR_PORT_STATS,
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_hw.h:170]: Are these STATS updates from David Ahern?  I know he's been patiently waiting to get some rocker device stats patches in.  I should just pull them in and update scottfeldman/qemu-rocker.

File: hw/net/rocker/rocker_hw.h [hunk 4]
```diff
@@ -165,6 +167,14 @@ enum {
...
+    ROCKER_TLV_CMD_TYPE_GET_PORT_STATS,
+
+    ROCKER_TLV_CMD_TYPE_P4_RMT_TABLE_ENTRY_ADD,
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_hw.h:173]: What does the "RMT" part of "P4_RMT" mean?

File: hw/net/rocker/rocker_hw.h [hunk 5]
```diff
@@ -475,6 +504,16 @@ enum rocker_of_dpa_overlay_type {
     (ROCKER_GROUP_TYPE_SET(ROCKER_OF_DPA_GROUP_TYPE_L3_UCAST) |\
      ROCKER_GROUP_INDEX_LONG_SET(index))
 
+/* cmd info (nested tlvs) for P4 */
+enum {
+    ROCKER_TLV_P4_RMT_INFO_UNSPEC = 0,  /* don't use 0 for tlv type */
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_hw.h:510]: Can probably drop the "INFO" part of the name to keep things shorter and be consistent with OF-DPA cmd info tlvs.

File: hw/net/rocker/rocker_world.c [hunk 6]
```diff
@@ -97,10 +108,8 @@ enum rocker_world_type world_type(World *world)
 
 const char *world_name(World *world)
 {
-    switch (world->type) {
-    case ROCKER_WORLD_TYPE_OF_DPA:
-        return "OF_DPA";
-    default:
-        return "unknown";
+    if (world) {
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_world.c]: Shouldn't need to check if world is non-null here.  If NULL was passed in, that's a bug in the calling code and I think we'd rather have the device crash (so the bug can get fixed) rather than returning something that sort of look like world is non-NULL.

Comment (scottfeldman) [inline hw/net/rocker/rocker_world.c]: I would say just return world->name here and that's it.

File: hw/net/rocker/rocker_world.c [hunk 7]
```diff
@@ -67,6 +75,9 @@ void world_free(World *world)
     if (world->ops->uninit) {
         world->ops->uninit(world);
     }
+    if (world->name) {
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_world.c]: Seems world->name will always be non-NULL here, so no need to check for NULL.

File: qapi/rocker.json [hunk 8]
```diff
@@ -75,7 +75,8 @@
 { 'type': 'RockerPort',
   'data': { 'name': 'str', 'enabled': 'bool', 'link-up': 'bool',
             'speed': 'uint32', 'duplex': 'RockerPortDuplex',
-            'autoneg': 'RockerPortAutoneg' } }
+            'autoneg': 'RockerPortAutoneg',
+            'world_name': 'str'} }
```
Comment (scottfeldman) [inline qapi/rocker.json]: nit: space between 'str' and '}'

File: Makefile [hunk 9]
```diff
@@ -290,7 +290,7 @@ clean:
 # avoid old build problems by removing potentially incorrect old files
 	rm -f config.mak op-i386.h opc-i386.h gen-op-i386.h op-arm.h opc-arm.h gen-op-arm.h
 	rm -f qemu-options.def
-	find . \( -name '*.l[oa]' -o -name '*.so' -o -name '*.dll' -o -name '*.mo' -o -name '*.[oda]' \) -type f -exec rm {} +
+	find . -path ./hw/net/rocker/p4lib -prune -o \( -name '*.l[oa]' -o -name '*.so' -o -name '*.dll' -o -name '*.mo' -o -name '*.[oda]' \) -type f -exec rm {} +
```
Comment (scottfeldman) [inline Makefile:293]: What does this do?  Doesn't seem like we want rocker stuff in the top-level Makefile

File: hw/net/Makefile.objs [hunk 10]
```diff
@@ -37,5 +37,6 @@ obj-$(CONFIG_ETSEC) += fsl_etsec/etsec.o fsl_etsec/registers.o \
 			fsl_etsec/rings.o fsl_etsec/miim.o
 
 common-obj-$(CONFIG_ROCKER) += rocker/rocker.o rocker/rocker_fp.o \
-                               rocker/rocker_desc.o rocker/rocker_world.o \
-                               rocker/rocker_of_dpa.o
+								rocker/rocker_desc.o rocker/rocker_world.o \
```
Comment (scottfeldman) [inline hw/net/Makefile.objs]: Strange tab issue here...not sure what changed

File: hw/net/rocker/README.p4 [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    rocker_p4_l2l3_tables.h
+    rocker_p4_l2l3_tables_if.h
+    rocker_p4_rmt.h
+    rocker_p4_features.h
+
+    p4lib/p4_l2l3_bm.a
+    p4lib/p4utils.a
+
+
+P4 behavioral model uses libJudy. Please install ligjudy-dev package.
```
Comment (scottfeldman) [inline hw/net/rocker/README.p4]: spelling ligjudy-dev

File: hw/net/rocker/README.p4 [hunk 12]
```diff
@@ -0,0 +1,45 @@
+P4 support for rocker device:
+--------------------------------
+
+P4 program is used define a new prcessing pipeline. This program
```
Comment (scottfeldman) [inline hw/net/rocker/README.p4]: spelling

File: hw/net/rocker/rocker.c [hunk 13]
```diff
@@ -452,20 +456,37 @@ static int cmd_consume(Rocker *r, DescInfo *info)
...
+        DPRINTF("Rocker port-get CMD %d\n", (int)cmd);
         err = cmd_get_port_settings(r, info, buf, info_tlv);
         break;
     case ROCKER_TLV_CMD_TYPE_SET_PORT_SETTINGS:
+        DPRINTF("Rocker port-set CMD %d\n", (int)cmd);
         err = cmd_set_port_settings(r, info_tlv);
         break;
     default:
-        err = -ROCKER_EINVAL;
+        if (tlvs[ROCKER_TLV_CMD_WORLD]) {
```
Comment (scottfeldman) [inline hw/net/rocker/rocker.c:498]: Hmm, let's just add the P4 cmds to case statement, and then world is known.

File: hw/net/rocker/rocker.c [hunk 14]
```diff
@@ -473,7 +494,7 @@ static void rocker_msix_irq(Rocker *r, unsigned vector)
 {
     PCIDevice *dev = PCI_DEVICE(r);
 
-    DPRINTF("MSI-X notify request for vector %d\n", vector);
+    // DPRINTF("MSI-X notify request for vector %d\n", vector);
```
Comment (scottfeldman) [inline hw/net/rocker/rocker.c]: leave uncommented

File: hw/net/rocker/rocker.c [hunk 15]
```diff
@@ -997,9 +1018,9 @@ static const char *rocker_reg_name(void *opaque, hwaddr addr)
 static void rocker_mmio_write(void *opaque, hwaddr addr, uint64_t val,
                               unsigned size)
 {
-    DPRINTF("Write %s addr " TARGET_FMT_plx
-            ", size %u, val " TARGET_FMT_plx "\n",
-            rocker_reg_name(opaque, addr), addr, size, val);
+    // DPRINTF("Write %s addr " TARGET_FMT_plx
```
Comment (scottfeldman) [inline hw/net/rocker/rocker.c]: leave uncommented

File: hw/net/rocker/rocker_fp.c [hunk 16]
```diff
@@ -119,6 +120,7 @@ int fp_port_eg(FpPort *port, const struct iovec *iov, int iovcnt)
     NetClientState *nc = qemu_get_queue(port->nic);
 
     if (port->enabled) {
+	    DPRINTF("Send to port %d\n", port->pport);
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_fp.c]: Might get too chatty?

File: hw/net/rocker/rocker_world.h [hunk 17]
```diff
@@ -56,5 +59,6 @@ enum rocker_world_type world_type(World *world);
 const char *world_name(World *world);
 
 World *rocker_get_world(Rocker *r, enum rocker_world_type type);
+World *rocker_world_from_pport(int pport);
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_world.h]: Is this used anywhere?

File: Makefile.target [hunk 18]
```diff
@@ -135,7 +135,7 @@ obj-y += memory.o savevm.o cputlb.o
 obj-y += memory_mapping.o
 obj-y += dump.o
 obj-$(call lnot,$(CONFIG_ROCKER)) += hw/net/rocker/qmp-norocker.o
-LIBS+=$(libs_softmmu)
+LIBS+=$(libs_softmmu) -lJudy
```
Comment (scottfeldman) [inline Makefile.target:138]: Is there something already like libjudy in the qemu build tree?  libjudy looks nice but I can see a road-block trying to get it included for upstream qemu.

File: hw/net/rocker/rocker_p4_l2l3_enums.h [hunk 19]
```diff
@@ -0,0 +1,41 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc. 
+
+Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_p4_l2l3_enums.h:4]: Can the generated files be licensed under GPL like the rest of the files?

File: hw/net/rocker/rocker_p4_l2l3_features.h [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#ifndef _P4_L2L3_FEATURES_H_
+#define _P4_L2L3_FEATURES_H_
+
+#define CPU_PORT  0xAA // b-model drops packet sent to ports 0, 257
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_p4_l2l3_features.h:19]: // style comments aren't allowed for qemu...need to convert to /\* */ comments.

File: hw/net/rocker/rocker_p4_l2l3_features.h [hunk 21]
```diff
@@ -0,0 +1,21 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_p4_l2l3_features.h:2]: Not sure about open-ended copyrights.  Past and current dates OK, but future?  I'm not sure...might raise a flag.

File: hw/net/rocker/rocker_p4_l2l3_lf.h [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+ */
+#ifndef _ROCKER_p4_l2l3_LF_H_
+#define _ROCKER_p4_l2l3_LF_H_
+typedef struct rocker_p4_l2l3_mac_learn_digest_digest_entry {
+  unsigned char ethernet_srcAddr[6];
+  unsigned short ingress_metadata_fid;
+  unsigned short standard_metadata_ingress_port;
+} rocker_p4_l2l3_mac_learn_digest_digest_entry_t;
+
+// Should be able to cast this to pipe_flow_lrn_msg_t.
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_p4_l2l3_lf.h:28]: // comment

File: hw/net/rocker/rocker_p4_l2l3_lf.h [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+typedef struct  rocker_p4_l2l3_mac_learn_digest_digest_msg {
+  unsigned int                  dev_tgt;
+  unsigned short                num_entries;
+  rocker_p4_l2l3_mac_learn_digest_digest_entry_t  *entries;
+} rocker_p4_l2l3_mac_learn_digest_digest_msg_t;
+
+// Should be able to cast this to pipe_flow_lrn_notify_cb.
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_p4_l2l3_lf.h:35]: // comment

File: hw/net/rocker/rocker_p4_rmt.c [hunk 24]
```diff
@@ -0,0 +1,216 @@
...
+    // ingress pipeline
+    DPRINTF("p4_rmt_ingress port %d, len %d .... \n", port, pkt_len);
+
+    // convert iov to a pkt buffer
+    unsigned char *buf = rocker_p4_iov_to_buffer(iov, iovcnt);
+    if (buf == NULL) {
+        return 0;
+    }
+    p4_rmt->process_pkt(port, buf, pkt_len);
+    return iov_size(iov, iovcnt);
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_p4_rmt.c]: return pkt_len?

File: hw/net/rocker/rocker_p4_rmt.c [hunk 25]
```diff
@@ -0,0 +1,216 @@
...
+{
+    int pkt_len = iov_size(iov, iovcnt);
+    unsigned int port = pport;  // model uses pports (1-N)
+    struct p4_rmt_world *p4_rmt = world_private(world);
+
+    // ingress pipeline
+    DPRINTF("p4_rmt_ingress port %d, len %d .... \n", port, pkt_len);
+
+    // convert iov to a pkt buffer
+    unsigned char *buf = rocker_p4_iov_to_buffer(iov, iovcnt);
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_p4_rmt.c]: I think in rocker (and this may be a qemu style thing) we tried to keep variable declarations before any code.

File: hw/net/rocker/rocker_p4_rmt.c [hunk 26]
```diff
@@ -0,0 +1,216 @@
...
+                            const struct iovec *iov, int iovcnt)
+{
+    int pkt_len = iov_size(iov, iovcnt);
+    unsigned int port = pport;  // model uses pports (1-N)
+    struct p4_rmt_world *p4_rmt = world_private(world);
+
+    // ingress pipeline
+    DPRINTF("p4_rmt_ingress port %d, len %d .... \n", port, pkt_len);
+
+    // convert iov to a pkt buffer
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_p4_rmt.c]: // comments need to be converted to /\* comments */

File: hw/net/rocker/rocker_p4_rmt.c [hunk 27]
```diff
@@ -0,0 +1,216 @@
...
+unsigned char *
+rocker_p4_iov_to_buffer(const struct iovec *iov, int iovcnt)
+{
+    // convert iov to a single contiguous buffer for p4 model
+    int len = 0;
+    unsigned char *buf = NULL;
+
+    len = iov_size(iov, iovcnt);
+    DPRINTF("convert %d iovs to a buffer of len %d\n", iovcnt, len);
+    buf = malloc(len);
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_p4_rmt.c]: Use g_malloc(), g_free()

File: hw/net/rocker/rocker_p4_rmt.c [hunk 28]
```diff
@@ -0,0 +1,216 @@
...
+
+    // ingress pipeline
+    DPRINTF("p4_rmt_ingress port %d, len %d .... \n", port, pkt_len);
+
+    // convert iov to a pkt buffer
+    unsigned char *buf = rocker_p4_iov_to_buffer(iov, iovcnt);
+    if (buf == NULL) {
+        return 0;
+    }
+    p4_rmt->process_pkt(port, buf, pkt_len);
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_p4_rmt.c]: not checking return code

File: hw/net/rocker/rocker_p4_rmt.c (hunk 24 above)
Comment (scottfeldman) [inline hw/net/rocker/rocker_p4_rmt.c]: Where does buf get freed?

File: hw/net/rocker/rocker_p4_rmt.c [hunk 29]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+static int rocker_p4_rmt_table_cmd (struct p4_rmt_world *p4_rmt, 
+                            struct desc_info *info,
+                            char *buf, unsigned short cmd,
+                            RockerTlv **tlvs)
+{
+    unsigned int    table_id;
+    unsigned char   *entry;
+    if (tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ID] == NULL ||
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_p4_rmt.c:114]: blank line after declarations but before fr executable code.

File: hw/net/rocker/rocker_p4_rmt.c [hunk 30]
```diff
@@ -0,0 +1,216 @@
...
+    int    iovcnt = 1;
+    struct rocker *rocker;
+    struct world *w;
+    unsigned int fp_ig_port;
+
+    struct p4_rmt_world *p4_rmt = (struct p4_rmt_world *)cookie; // world_private(w);
+    rocker = p4_rmt->rocker;
+    w = p4_rmt->world;
+
+    fp_port_from_pport((unsigned int)ig_port1, &fp_ig_port);
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_p4_rmt.c:76]: where is fp_ig_port used?

File: hw/net/rocker/rocker_p4_rmt.c [hunk 31]
```diff
@@ -0,0 +1,216 @@
...
+    if (!p4_rmt->is_table_valid(table_id)) {
+        DPRINTF("p4_rmt_table_cmd for invalide table %d\n", table_id);
+        return 0;
+    }
+
+    switch (cmd) {
+        case ROCKER_TLV_CMD_TYPE_P4_RMT_TABLE_ENTRY_ADD:
+            // check if fn is provided
+            if (p4_rmt->table_ops[table_id].add) {
+                p4_rmt->table_ops[table_id].add(entry);
```
Comment (scottfeldman) [inline hw/net/rocker/rocker_p4_rmt.c:135]: Hmm...it looks like entry is a binary blob.  How is it packed?  Are there un-aligned access issues?  How is it versioned?  How is big/little/host/network byte order handled?  It seems we want TLV-style encode/decode of the entry fields to answer all the questions.

File: hw/net/rocker/rocker_hw.h (hunk 2 above)
Comment (pbhide) [inline hw/net/rocker/rocker_hw.h:150]: Yes, currently the cmds are unique across the worlds (but don't have to
be). I started without the world_id tlv but introduced the world_id tlv for
the following two reasons. I also took the cue from the comment just above
the switch in cmd_consume() function. -
1. When a new cmd for any world is added, I had to add it in two places the
   switch in cmd_consume() function in rocker.c and again in world-specific
   command handler (e.g of_dpa_cmd() or rocker_p4_rmt_cmd())
2. As we add more P4 pipelines (worlds), we plan to provide an uniform
   interface for all table operations (CRUD). So new table specific TLVs are
   not visible in rocker_rmt_cmd either).

Basically, when something changes in a world (pipeline), I wanted to
minimize code changes to the common code in rocker.c and rocker_p4_rmt.c
(this file is common to different P4 worlds). With world_id tlv, I removed
changes in rocker.c and with a table_op blob, I do not need to change the
rocer_p4_rmt.c
Also as we add more worlds they can have there own TLV space (like nested
TLVs, but w/o adding another level of nesting).

File: hw/net/rocker/rocker_hw.h (hunk 3 above)
Comment (pbhide) [inline hw/net/rocker/rocker_hw.h:170]: Yes, I copied them from the driver code when I updated net-next. Ran into
tlv types not matching between driver and device. I think once all repos
are up to date this will not show up as a diff.

File: hw/net/rocker/rocker_hw.h (hunk 4 above)
Comment (pbhide) [inline hw/net/rocker/rocker_hw.h:173]: RMT = Reconfigurable Match Table  - This achronym comes from one of the
initial papers on reconfigurable packet processing (I'll send you the exact
name of the paper).
I am using it here as a common identifier that can be shared by P4 worlds
(L2L3 and future). Each P4 world will add its own tables but share a single
TLV type to ADD and table entry.

File: hw/net/rocker/rocker_hw.h (hunk 5 above)
Comment (pbhide) [inline hw/net/rocker/rocker_hw.h:510]: I wanted to have something in TLV that tells me that this is a nested TLV
(so it won't be used incorrectly).
I think, I wanted to drop INFO and keep CMD but did the reverse by mistake.
E.g. I wanted -
ROCKER_TLV_P4_RMT_CMD_UNSPEC, ..
This will be consistent with -
ROCKER_TLV_CMD_PORT_SETTINGS_UNSPEC

Eitherways.. it really does not shorten the name. I can drop both INFO and
CMD (altough prefer to keep some indication that this is nested tlv under
CMD)

File: hw/net/rocker/rocker_world.c (hunk 6 above)
Comment (pbhide) [inline hw/net/rocker/rocker_world.c]: Agree. I really wanted as assert.. but don't see it being used anywhere.
Also wanted to somewhat mimic the default: case in the earlier code.

Removing the check.

File: hw/net/rocker/rocker_world.c (hunk 7 above)
Comment (pbhide) [inline hw/net/rocker/rocker_world.c]: yes. will remove the check.

File: Makefile (hunk 9 above)
Comment (pbhide) [inline Makefile:293]: Make clean at the top level was removing all libraries in all directories.
Since P4libs are currently not built as part of emu build they should not
be removed. These are checked-in libraries until we find a way to compiler
P4 program as part of emu build.
I added the above so that make clean does not remove the P4 libs. Could not
find more elegant way.

File: hw/net/Makefile.objs (hunk 10 above)
Comment (pbhide) [inline hw/net/Makefile.objs]: My bad.. I had tab-stop set to 4 spaces.

File: hw/net/rocker/rocker.c (hunk 13 above)
Comment (pbhide) [inline hw/net/rocker/rocker.c:498]: Same reasoning as earlier comment.

File: hw/net/rocker/rocker_world.h (hunk 17 above)
Comment (pbhide) [inline hw/net/rocker/rocker_world.h]: not anymore. will remove.

File: Makefile.target (hunk 18 above)
Comment (pbhide) [inline Makefile.target:138]: libjudy is used in the P4 behavioral model. That might be a big change for
us to migrate to something else like GHash.
Do you think it can result in "no-go"?

File: hw/net/rocker/rocker_p4_l2l3_enums.h (hunk 19 above)
Comment (pbhide) [inline hw/net/rocker/rocker_p4_l2l3_enums.h:4]: Its a generated file.. let me try to fix the tool that generates it.

File: hw/net/rocker/rocker_p4_l2l3_features.h (hunk 21 above)
Comment (pbhide) [inline hw/net/rocker/rocker_p4_l2l3_features.h:2]: I had the same question. I think it has to be a specific year. I'll work
internally to get it fixed.

File: hw/net/rocker/rocker_p4_rmt.c (hunk 27 above)
Comment (pbhide) [inline hw/net/rocker/rocker_p4_rmt.c]: This was done on purpose.. (But I need to fix it) as the P4 behavior model
frees the packet buffer using "free" not g_free.
But the model code does not allocate the buffer.. I'll abstract it by
creating pkt_buf_alloc() function that will be provided by the model
library.

File: hw/net/rocker/rocker_p4_rmt.c (hunk 28 above)
Comment (pbhide) [inline hw/net/rocker/rocker_p4_rmt.c]: Actually the process_pkt (currently) always returns 0. Also was not sure
what to return on the error.

How is the return value of the  NetClientInfo->receive_iov used by its
caller?
Looks like the receive_iov should return the bytes consumed ?? Can I return
0 if packet could not be processed for some reason.. Need to check.

I'll add the check.

File: hw/net/rocker/rocker_p4_rmt.c (hunk 24 above)
Comment (pbhide) [inline hw/net/rocker/rocker_p4_rmt.c]: By the model code. (library)

File: hw/net/rocker/rocker_p4_rmt.c (hunk 31 above)
Comment (pbhide) [inline hw/net/rocker/rocker_p4_rmt.c:135]: Yes it is a blob. One good thing is that, the code to create and consume
the entry  is generated by the p4 tool-chain so bit packing and endian-ness
is taken care of by the generated code. E.g. rocker_p4_l2l3_pd.c which has
entry build functions.

File: Makefile.target (hunk 18 above)
Comment (scottfeldman) [inline Makefile.target:138]: If the answer to "can ghash be used here instead of libjudy" is yes, then you probably need to convert to ghash.  I would expect push-back from qemu community on adding libjudy and this is the first question they'll ask.

File: hw/net/rocker/rocker_hw.h (hunk 2 above)
Comment (scottfeldman) [inline hw/net/rocker/rocker_hw.h:150]: Ok, those reasons make sense.  I guess OF-DPA is the exception.  We might be able to promote of-dpa to this model without breaking backward-compat with older drivers by assuming if ROCKER_TLV_CMD_WORLD is not defined in the tlvs then it must be an of-dpa cmd.

File: hw/net/rocker/rocker_hw.h (hunk 3 above)
Comment (scottfeldman) [inline hw/net/rocker/rocker_hw.h:170]: makes sense

File: hw/net/rocker/rocker_hw.h (hunk 4 above)
Comment (scottfeldman) [inline hw/net/rocker/rocker_hw.h:173]: I see.  Maybe there is a place for a comment in the code that expands the acronym?

File: hw/net/rocker/rocker_hw.h (hunk 5 above)
Comment (scottfeldman) [inline hw/net/rocker/rocker_hw.h:510]: Ok, maybe INFO is best since that's the nest it's under.  Again, of-dpa will be an exception, but maybe we can fix of-dpa to be consistent once the dust settles.

File: hw/net/rocker/rocker_p4_rmt.c (hunk 27 above)
Comment (scottfeldman) [inline hw/net/rocker/rocker_p4_rmt.c]: ok, cool

File: hw/net/rocker/rocker_p4_rmt.c (hunk 31 above)
Comment (scottfeldman) [inline hw/net/rocker/rocker_p4_rmt.c:135]: If it's generated code, then can a TLV format be generated?  I know that brings rocker's TLV packing into P4 tools, which is probably not what you want.  I think a blob is not going to be the correct answer, but I'm not sure where to slice this.  This is something we should revisit after all the other review work is done.
